agent/lib_mongodb.c

```diff
@@ -173,7 +175,162 @@ NR_PHP_WRAPPER(nr_mongodb_operation) {
 }
 NR_PHP_WRAPPER_END
 
-void nr_mongodb_enable(TSRMLS_D) {
+#else
+
+NR_PHP_WRAPPER(nr_mongodb_operation_before) {
+  (void)wraprec;
+  nr_segment_t* segment;
+  segment = nr_segment_start(NRPRG(txn), NULL, NULL);
```

Comment:
If [this check](https://github.com/newrelic/newrelic-php-agent/blob/64c6306ea27dc9f8474daae940f8fd63054ab2c6/agent/lib_mongodb.c#L211-L222) would be performed in `nr_mongodb_operation_before` and would fail, `nr_mongodb_operation_before` wouldn't create a new segment that is discarded in `nr_mongodb_operation_after`. I don't know which approach results in smaller agent's overhead. Has this been evaluated and therefore [this check](https://github.com/newrelic/newrelic-php-agent/blob/64c6306ea27dc9f8474daae940f8fd63054ab2c6/agent/lib_mongodb.c#L211-L222) is only done in `nr_mongodb_operation_after`?

Reply:
These blocks of code are largely recycled from the non-OAPI implementation and modified to suit the OAPI paradigm.
- https://github.com/newrelic/newrelic-php-agent/blob/64c6306ea27dc9f8474daae940f8fd63054ab2c6/agent/lib_mongodb.c#L113
> Has this been evaluated and therefore [this check](https://github.com/newrelic/newrelic-php-agent/blob/64c6306ea27dc9f8474daae940f8fd63054ab2c6/agent/lib_mongodb.c#L211-L222) is only done in nr_mongodb_operation_after?

I haven't evaluated performance implications of this. I doubt we have the capability to do so to the granular degree that would be required here with confidence. Are you suggesting to duplicate the check and have it run not only in before but after as well? We would also have to handle not initializing the segment in both before in a way that after could be aware of it, then modify its behavior accordingly.

I don't believe that this case gets hit very often. There's no historical explanation or conversation over why it exists that I can find, and it may be that it should just be removed. All the [functions](https://github.com/mongodb/mongo-php-library/tree/master/src/Operation) we wrap should satisfy the `MongoDB\\Operation\\Executable` interface check, so I'm not sure if this actually ever gets triggered or if it was a relic of the past.

Reply:
I guess the likelihood of the test failing is small. I'm ok with leaving the code as is.